File: converted_code/backend/src/main/java/com/zeriq/ai/controller/CourseController.java
package com.zeriq.ai.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import com.zeriq.ai.model.Course;
import com.zeriq.ai.model.Lesson;
import com.zeriq.ai.model.Resource;
import com.zeriq.ai.repository.CourseRepository;
import com.zeriq.ai.repository.LessonRepository;
import com.zeriq.ai.repository.ResourceRepository;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/courses")
@CrossOrigin(origins = "*")
public class CourseController {

    @Autowired
    private CourseRepository courseRepository;

    @Autowired
    private LessonRepository lessonRepository;

    @Autowired
    private ResourceRepository resourceRepository;

    @GetMapping
    public List<Course> getAllCourses() {
        return courseRepository.findAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getCourseById(@PathVariable Long id) {
        return courseRepository.findById(id)
                .<ResponseEntity<?>>map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("error", "Course not found.")));
    }

    @GetMapping("/{courseId}/lessons")
    public List<Lesson> getLessonsByCourse(@PathVariable Long courseId) {
        return lessonRepository.findByCourseIdOrderByOrderIndexAsc(courseId);
    }

    @GetMapping("/lessons/{lessonId}")
    public ResponseEntity<?> getLessonById(@PathVariable Long lessonId) {
        return lessonRepository.findById(lessonId)
                .<ResponseEntity<?>>map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("error", "Lesson not found.")));
    }

    @GetMapping("/{courseId}/resources")
    public List<Resource> getResourcesByCourse(@PathVariable Long courseId) {
        return resourceRepository.findByCourseId(courseId);
    }
}

File: src/backend/controllers/CourseController.ts
import { Request, Response, Router } from "express";
import { CourseService } from "../services/CourseService";

export class CourseController {
  private courseService = new CourseService();
  public router = Router();

  constructor() {
    this.router.get("/", this.getCourses.bind(this));
    this.router.get("/:id", this.getCourse.bind(this));
    this.router.get("/:id/lessons", this.getLessons.bind(this));
    this.router.get("/lessons/:lessonId", this.getLesson.bind(this));
    this.router.get("/:id/resources", this.getResources.bind(this));
  }

  private getCourses(req: Request, res: Response): void {
    try {
      const courses = this.courseService.getCourses();
      res.json(courses);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private getCourse(req: Request, res: Response): void {
    try {
      const course = this.courseService.getCourseById(Number(req.params.id));
      if (!course) {
        res.status(404).json({ error: "Course not found" });
        return;
      }
      res.json(course);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private getLessons(req: Request, res: Response): void {
    try {
      const lessons = this.courseService.getLessonsByCourse(Number(req.params.id));
      res.json(lessons);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private getLesson(req: Request, res: Response): void {
    try {
      const lesson = this.courseService.getLessonById(Number(req.params.lessonId));
      if (!lesson) {
        res.status(404).json({ error: "Lesson not found" });
        return;
      }
      res.json(lesson);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }

  private getResources(req: Request, res: Response): void {
    try {
      const resources = this.courseService.getResourcesByCourse(Number(req.params.id));
      res.json(resources);
    } catch (e: any) {
      res.status(500).json({ error: e.message });
    }
  }
}
